core/src/main/java/io/questdb/griffin/engine/groupby/GroupByUtils.java

```diff
@@ -218,6 +219,9 @@ public static void prepareGroupByRecordFunctions(
                     functionKeyColumnIndex++;
                     func = createColumnFunction(null, functionKeyColumnIndex, type, -1);
                     keyTypes.add(functionKeyColumnIndex - valueCount - 1, func.getType());
```

Comment:
@puzpuzpuz  Please have a look at this change.

Reply:
I plan to work on parallel `count_distinct`, `min(str)`/`max(str)`, and other similar aggregate functions once I'm back from the hols. This should speed up all keyed queries that involve `count_distinct` - considering this, I'm not sure if we need this SQL rewrite. WDYT?

Off-topic: I've created #4085 as a potential follow-up for parallel GROUP BY. It should help a lot in a number of ClickBench queries.

Reply:
Feel free to close the if it's not needed.
It'd be good to fix the issue mentioned above (not necesarily this way) because it breaks e.g. when Async Group By   is wrapped by VirtualRecordCursor.

Reply:
Which issue do you mean? I don't see any references in the PR description.

Reply:
The issue is triggered by:
```sql
create table x as (select * from (select rnd_symbol('a','b','c') a, 'x' || x b from long_sequence(20)))

select a, ct, 1
from 
(
  select a, cast(b as SYMBOL) as ct , count(*)
  from x
)
```

Stack trace:
java.lang.ClassCastException: class io.questdb.griffin.engine.functions.columns.StrColumn cannot be cast to class io.questdb.cairo.sql.SymbolTable (io.questdb.griffin.engine.functions.columns.StrColumn and io.questdb.cairo.sql.SymbolTable are in module io.questdb of loader 'app')
	at io.questdb.griffin.engine.table.AsyncGroupByRecordCursor.getSymbolTable(AsyncGroupByRecordCursor.java:127)
	at io.questdb.griffin.engine.functions.columns.SymbolColumn.init(SymbolColumn.java:91)
	at io.questdb.cairo.sql.Function.init(Function.java:48)
	at io.questdb.griffin.engine.table.VirtualRecordCursorFactory.getCursor(VirtualRecordCursorFactory.java:83)

Reply:
Thanks for the clarification. I hope that #4097 turns out as I expect and we land it, but in any case, I'll make sure to fix this bug. Let's keep your PR open for now and decide what to do with it a bit later.